### ISSUE TITLE
Unwanted Characters in Console Output

### ISSUE THREAD

**Issue by Burtannia** [commercialhaskell/stack#4091]
Hi,

I'm getting extra characters appearing in my console output. After doing some research I discovered that there were some issues with colour codes and `ansi-terminal` on Windows. From the bug reports I read that they had been fixed in a recent commit. I therefore updated Stack to the Git version, `stack --version` now gives the following:
`Version 1.8.0, Git revision faa941a2d697b252a2f69e8ddd3f276ff44bf8b6 x86_64 hpack-0.28.2`

However, the characters still appear after the update. Adding `--color never` fixes the issue as mentioned in a [previous bug report](https://github.com/commercialhaskell/stack/issues/3797). The characters are seen both in the terminal within VS Code and the Windows Command Prompt. The characters do not appear in the terminal or within VS Code on Ubuntu.

`←[0mAccounting-0.1.0.0: unregistering (local file changes: app\Main.hs)←[0m
←[0mAccounting-0.1.0.0: build (lib + exe)←[0m
←[0mPreprocessing library for Accounting-0.1.0.0..←[0m
←[0mBuilding library for Accounting-0.1.0.0..←[0m
←[0mPreprocessing executable 'Accounting-exe' for Accounting-0.1.0.0..←[0m
←[0mBuilding executable 'Accounting-exe' for Accounting-0.1.0.0..←[0m
←[0m[2 of 2] Compiling Main             ( app\Main.hs, .stack-work\dist\5c8418a7\build\Accounting-exe\Accounting-exe-tmp\Main.o )←[0m
←[0mLinking .stack-work\dist\5c8418a7\build\Accounting-exe\Accounting-exe.exe ...←[0m
←[0mAccounting-0.1.0.0: copy/register←[0m
←[0mInstalling library in C:\Projects\Accounting\.stack-work\install\e6d3037a\lib\x86_64-windows-ghc-8.2.2\Accounting-0.1.0.0-F8ZttcOqREWElkyQQVED0s←[0m
←[0mInstalling executable Accounting-exe in C:\Projects\Accounting\.stack-work\install\e6d3037a\bin←[0m
←[0mRegistering library for Accounting-0.1.0.0..←[0m`

Stack version: 1.8.0
LTS: 11.13
Global GHC version: 8.0.2
OS: Windows 7 Professional 64-bit

**Comment by snoyberg** [comment]
FWIW, I'm unable to reproduce this on my Windows 10 machine.

**Comment by Burtannia** [comment]
@snoyberg Can confirm that this issue doesn't arise with the Git version of Stack on Windows 10. This is probably not worth looking into if it's only Windows 7.

**Comment by mpilgrem** [comment]
The native terminals in Windows 7 are not ANSI-capable, so emulation is required for ANSI-like behaviour. Package `ansi-terminal` provides that emulation, but `stack` no longer uses that library for all of its ANSI output (it still uses it for some of its output, I believe). `stack`, however, does use `hSupportsANSI` from the package, which seeks to detect if `ansi-terminal` emulation would work (if, and to the extent, `stack` uses it).

**Comment by snoyberg** [comment]
My original thought had been to default Windows to no color output. Another option would be to do this for Windows < 10. Maybe that's a safer option.

**Comment by mpilgrem** [comment]
@snoyberg Non-native terminals on legacy Windows that are ANSI-capable are freely available (and the current `--git` version of `stack` works with them), so I would suggest that disabling `stack` colour functionality on Windows (or even only on legacy Windows) by default would not be the best course of action. That is, the issue is terminal-related not OS-related, although native terminals and their OS go hand in hand.

Perhaps it would be better to provide (online or in the application itself) clearer and more comprehensive help to potential users of `stack` on legacy Windows using native terminals, explaining that `stack` does not fully support them and the work-arounds available (eg use a non-native terminal or  disable some `stack` functionality with `--color=never` - I am assuming that can be set in `config.yaml` ).

**Comment by snoyberg** [comment]
This is becoming a recurring source of bug reports, which is the kind of thing I try to avoid. I can't personally imagine any set of clearer help or docs that will resolve this situation beyond turning off color for Windows < 10.

**Comment by mpilgrem** [comment]
Currently, `stack --help` prompts:
````
--color WHEN             Specify when to use color in output; WHEN is
                         'always', 'never', or 'auto'
````
Perhaps it would help some users if it were extended to say:
````
--color WHEN             Specify when to use color in output; WHEN is
                         'always', 'never', or 'auto'. NB: on Windows
                         versions before Windows 10, color codes will
                         corrupt some output in Command Prompt or
                         PowerShell terminals.
````

**Comment by snoyberg** [comment]
I highly doubt it. @Burtannia you just hit this issue yourself, would you care to weigh in?

**Comment by Burtannia** [comment]
@snoyberg @mpilgrem I'm not sure that it's particularly intuitive for users to refer to `stack --help` when encountering a problem like this. When I first saw the characters I had no idea that they were to do with colour codes let alone that there was an option to turn off colour.

I searched around a bit and found the other bug reports that mentioned `--color never` then I found the mention that a commit had fixed the issue. After seeing that the commit was more recent that the latest release I ran `stack upgrade --git` only to find the issue remained hence why I made this report.

Hopefully that provides some insight as to the line of thinking a user might have although I can only speak for how I reacted.

I don't see a particular disadvantage to setting `--color never` by default on < Windows 10. If a user would like to add the colour in the case they're using a different terminal then that might be a more intuitive scenario to use `stack --help` where there could be a message regarding enabling it.

**Comment by mpilgrem** [comment]
If that is the way forward, then it is probably better to:

* on Windows, test for the absence of the feature (ie `ENABLE_VIRTUAL_TERMINAL_PROCESSING` not succeeding - https://docs.microsoft.com/en-us/windows/console/setconsolemode) than for specific versions of Windows (given the advice here: https://msdn.microsoft.com/en-us/library/windows/desktop/ms724832(v=vs.85).aspx); and

* perhaps, provide an enhanced prompt like:
````
--color WHEN             Specify when to use color in output; WHEN is
                         'always', 'never', or 'auto'. On Windows
                         versions before Windows 10, the default is
                         'never' but color may work on terminals
                         that support color codes.

**Comment by dbaynard** [comment]
It looks like there's agreement on the way to proceed on this — correct me if I'm wrong — and that the best resolution would be to disable colour when the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` feature is not possible.

This is probably suitable for a new contributor. Given the comments in #3992, though, it would be a good idea to resolve this quickly. So @Burtannia and @mpilgrem, would you be willing to test the changes, or even make a PR?

**Comment by mpilgrem** [comment]
I will try to provide a pull request soon.

**Comment by Burtannia** [comment]
I was looking up how to check `ENABLE_VIRTUAL_TERMINAL_PROCESSING` and I found the code for `ansi-terminal`. Doesn't `stack` already use `hSupportsANSI` to test this? Unless I'm missing something (which I probably am) I don't see how testing `ENABLE_VIRTUAL_TERMINAL_PROCESSING` again inside `stack` will produce a different result.

**Comment by mpilgrem** [comment]
`hSupportsANSI` tests if a handle will work with the `ansi-terminal` library. It does not test if a handle will support ANSI codes without the library's ability to fall back on emulation. My pull request - hopefully coming in a few hours; my access to a Windows 7 machine for testing is limited - duplicates the `ansi-terminal` internals. The test `ansi-terminal` uses to decide to emulate can be re-purposed for `stack` to decide (by default) not to support colour.

**Comment by Burtannia** [comment]
Ok that makes sense. I mainly run Windows 7 so I can check the fix once it's on Git.

**Comment by Burtannia** [comment]
@mpilgrem I've tested your commit and everything seems to work.

I get no colour and no extra characters when using the command line or powershell.
Colours appear as expected when using Cygwin (whereas with the old version of `stack` colour wasn't displayed).

**Comment by dbaynard** [comment]
Good work everyone! I'll be commenting further over on the PR so @Burtannia you may wish to follow it.

**Comment by snoyberg** [comment]
I've merged #4106 (thanks @mpilgrem!). Closing this issue, if the problem still appears on master @Burtannia, please add another comment and I'll reopen.

**Comment by Topsii** [comment]
@snoyberg The problem still appears for me, but I am not sure in where you want to track it. 
See my comment: https://github.com/commercialhaskell/stack/issues/3797#issuecomment-425799747

**Comment by snoyberg** [comment]
Have you tested the recent 1.9 release candidate?

**Comment by mpilgrem** [comment]
@snoyberg, the commit that introduced the bug is post-1.9. See the proposed fix at #4332.
